internal/uvm/virtual_device.go

```diff
@@ -14,55 +14,104 @@ import (
 // VPCIDevice represents a vpci device. Holds its guid and a handle to the uvm it
 // belongs to.
 type VPCIDevice struct {
-	vm *UtilityVM
-	ID string
+	vm               *UtilityVM
+	VMBusGUID        string
+	deviceInstanceID string
+	refCount         uint32
 }
 
 // Release frees the resources of the corresponding vpci device
 func (vpci *VPCIDevice) Release(ctx context.Context) error {
-	if err := vpci.vm.RemoveDevice(ctx, vpci.ID); err != nil {
+	if err := vpci.vm.RemoveDeviceFromUVM(ctx, vpci); err != nil {
 		log.G(ctx).WithError(err).Warn("failed to remove VPCI device")
 		return err
 	}
 	return nil
 }
 
-// AssignDevice assigns a new vpci device to the uvm
-func (uvm *UtilityVM) AssignDevice(ctx context.Context, device hcsschema.VirtualPciDevice) (*VPCIDevice, error) {
+// AssignDeviceToUVM assigns a vpci device to the uvm
+// if the device already exists, the stored VPCIDevice's ref count is increased
+// and the VPCIDevice is returned.
+// Otherwise, a new request is made to assign the target device indicated by the deviceID
+// onto the UVM. A new VPCIDevice entry is made on the UVM and the VPCIDevice is returned
+// to the caller
+func (uvm *UtilityVM) AssignDeviceToUVM(ctx context.Context, deviceID string) (*VPCIDevice, error) {
 	guid, err := guid.NewV4()
 	if err != nil {
 		return nil, err
 	}
-	id := guid.String()
+	vmBusGUID := guid.String()
 
 	uvm.m.Lock()
 	defer uvm.m.Unlock()
-	if err := uvm.modify(ctx, &hcsschema.ModifySettingRequest{
-		ResourcePath: fmt.Sprintf(virtualPciResourceFormat, id),
+
+	if uvm.vpciIDs == nil {
+		uvm.vpciIDs = make(map[string]*VPCIDevice)
+	}
+
+	existingVPCIDevice := uvm.vpciIDs[deviceID]
+	if existingVPCIDevice != nil {
+		existingVPCIDevice.refCount++
+		return existingVPCIDevice, nil
+	}
+
+	targetDevice := hcsschema.VirtualPciDevice{
+		Functions: []hcsschema.VirtualPciFunction{
+			{
+				DeviceInstancePath: deviceID,
+			},
+		},
+	}
+
+	request := &hcsschema.ModifySettingRequest{
+		ResourcePath: fmt.Sprintf(virtualPciResourceFormat, vmBusGUID),
 		RequestType:  requesttype.Add,
-		Settings:     device,
-		GuestRequest: guestrequest.GuestRequest{
+		Settings:     targetDevice}
+
+	if uvm.operatingSystem != "windows" {
+		// opengcs call to make sure device is present in the uvm
+		guestRequest := guestrequest.GuestRequest{
```

Comment:
Why not just 'request.GuestRequest = guestrequest.GuestRequest{...}'

Reply:
Just thought this was clearer for me. Would you prefer I change it?

Reply:
I think its a matter of preference but it gets rid of an entry in the guestRequest/GuestRequest/guestrequest hierarchy hahaha. I'd prefer it though yes